File: packages/sagar_core/geo.py
"""Coordinate handling, CRS constants and geodesic helpers.

No GDAL, no shapely, no pyproj. Those live in the geo/micromamba
environment and belong to the packages that actually do geometry work
(sagar_sar, sagar_drift). What lives here is the small set of primitives
that every package — including the API, which runs in a different
environment entirely — needs to agree on: what a coordinate is, which
way round it goes, and how far apart two of them are.
"""

from __future__ import annotations

import math
from typing import NamedTuple

from .units import METRES_PER_NAUTICAL_MILE

#: WGS84. Every coordinate in this system is EPSG:4326 unless a module
#: says otherwise in its own docstring, and none currently do.
CRS_WGS84 = "EPSG:4326"

#: Mean Earth radius (IUGG), metres. Used for haversine distance.
EARTH_RADIUS_M = 6_371_008.8


class LonLat(NamedTuple):
    """A WGS84 coordinate, longitude first.

    Longitude first because that is GeoJSON's order (RFC 7946 §3.1.1) and
    the API contract's stated convention, and being consistent with the
    wire format everywhere is worth more than matching the "lat, lon"
    order humans say out loud.

    A NamedTuple rather than a Pydantic model on purpose: these are
    constructed in tight loops (per-particle, per-AIS-ping) where model
    validation overhead is real. Validation happens at boundaries via
    `validate_lonlat`, not on every construction.
    """

    lon: float
    lat: float


def validate_lonlat(lon: float, lat: float) -> LonLat:
    """Range-check a coordinate and return it.

    Catches a swapped pair ONLY when the true longitude exceeds ±90, so
    the swapped value falls outside valid latitude range. Within our own
    Arabian Sea AOI (roughly 68–78°E, 8–15°N) both components are
    individually valid latitudes, so a swap there passes silently — this
    is a genuine, unavoidable gap in any pure range check, documented
    here and mirrored in services/api/app/schemas.py's LonLat. Use
    `in_bbox` against a known AOI when you need to catch that case.
    """
    if not (-180.0 <= lon <= 180.0):
        raise ValueError(f"longitude out of range [-180, 180]: {lon}")
    if not (-90.0 <= lat <= 90.0):
        raise ValueError(
            f"latitude out of range [-90, 90]: {lat} — "
            "this usually means the pair was written [lat, lon] by mistake"
        )
    return LonLat(lon, lat)


class BBox(NamedTuple):
    """An axis-aligned bounding box, WGS84 degrees.

    Does not handle antimeridian crossing (min_lon > max_lon). Our AOIs
    are the Arabian Sea and Bay of Bengal, neither of which comes near
    180°, so supporting it would be untested code guarding against a case
    that cannot currently arise. `from_bounds` raises rather than silently
    producing a box that spans the wrong 340 degrees.
    """

    min_lon: float
    min_lat: float
    max_lon: float
    max_lat: float

    @classmethod
    def from_bounds(cls, min_lon: float, min_lat: float, max_lon: float, max_lat: float) -> BBox:
        if min_lon > max_lon:
            raise ValueError(
                f"min_lon {min_lon} > max_lon {max_lon}: antimeridian-crossing boxes "
                "are not supported (and no SAGARDRISHTI AOI needs one)"
            )
        if min_lat > max_lat:
            raise ValueError(f"min_lat {min_lat} > max_lat {max_lat}")
        validate_lonlat(min_lon, min_lat)
        validate_lonlat(max_lon, max_lat)
        return cls(min_lon, min_lat, max_lon, max_lat)

    @classmethod
    def parse(cls, text: str) -> BBox:
        """Parse the `?bbox=minlon,minlat,maxlon,maxlat` query form used
        by the API contract."""
        parts = [p.strip() for p in text.split(",")]
        if len(parts) != 4:
            raise ValueError(
                f"bbox must be 'minlon,minlat,maxlon,maxlat' (4 values), got {len(parts)}: {text!r}"
            )
        try:
            nums = [float(p) for p in parts]
        except ValueError as exc:
            raise ValueError(f"bbox components must be numbers: {text!r}") from exc
        return cls.from_bounds(*nums)


def in_bbox(point: LonLat, bbox: BBox) -> bool:
    return bbox.min_lon <= point.lon <= bbox.max_lon and bbox.min_lat <= point.lat <= bbox.max_lat


def haversine_m(a: LonLat, b: LonLat) -> float:
    """Great-circle distance in metres.

    Haversine on a spherical Earth: sub-0.5% error against the WGS84
    ellipsoid, which is far below the ~8 km resolution of the ocean
    current fields that dominate our error budget (docs/PRD.md §6.5). A
    geodesic solution would be more precise and would need pyproj, which
    this package deliberately does not depend on.
    """
    lat1, lat2 = math.radians(a.lat), math.radians(b.lat)
    dlat = lat2 - lat1
    dlon = math.radians(b.lon - a.lon)
    h = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
    return 2 * EARTH_RADIUS_M * math.asin(math.sqrt(h))


def haversine_nm(a: LonLat, b: LonLat) -> float:
    """Distance in nautical miles — the unit maritime work actually uses,
    and the one the AIS coverage threshold (~40–75 nm) is expressed in."""
    return haversine_m(a, b) / METRES_PER_NAUTICAL_MILE


def initial_bearing_deg(a: LonLat, b: LonLat) -> float:
    """Initial great-circle bearing from `a` to `b`, degrees in [0, 360).

    "Initial" matters: along a great circle the bearing changes
    continuously, so this is the heading at `a`, not an average. Over the
    short hops between consecutive AIS pings the difference is
    negligible; over a long track segment it is not.
    """
    lat1, lat2 = math.radians(a.lat), math.radians(b.lat)
    dlon = math.radians(b.lon - a.lon)
    x = math.sin(dlon) * math.cos(lat2)
    y = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(dlon)
    return math.degrees(math.atan2(x, y)) % 360.0


def implied_speed_knots(a: LonLat, b: LonLat, seconds: float) -> float:
    """Speed implied by moving from `a` to `b` in `seconds`.

    This is the backbone of the AIS "teleport" data-quality check
    (docs/SCORING_MODEL.md §2.1(a)): a vessel reporting positions 50 nm
    apart 30 seconds apart did not travel at 6000 knots, it emitted bad
    data, and that record must be marked unreliable rather than fed to
    the scoring model.

    Raises on non-positive elapsed time — two positions with the same
    timestamp imply an infinite speed, which is a data defect to surface,
    not a float('inf') to propagate into a suspicion score.
    """
    if seconds <= 0:
        raise ValueError(f"elapsed seconds must be positive, got {seconds}")
    return haversine_nm(a, b) / (seconds / 3600.0)

File: packages/go/geo/geo.go
// Package geo is the small set of geodesic primitives the Go side of the
// AIS data plane needs: distance and bearing between two points, and a
// point some fraction of the way along a path. See ADR 0005 - this and
// aoi/quality/store are what services/aisgen and services/aisd share, so
// synthetic and recorded traffic use the same math, not two copies of it.
//
// Deliberately not a general geospatial library: no polygon ops, no
// projections, nothing PostGIS-shaped. Anything more than the three
// functions below belongs in a PostGIS query or in Python (GeoPandas,
// shapely) per ADR 0005's boundary, not here.
package geo

import "math"

// EarthRadiusKm is the mean radius used throughout this package. Good to
// well under 1% at the distances a shipping lane covers - nobody needs a
// WGS84 ellipsoid for "is this vessel near this lane".
const EarthRadiusKm = 6371.0088

func toRad(deg float64) float64 { return deg * math.Pi / 180 }
func toDeg(rad float64) float64 { return rad * 180 / math.Pi }

// HaversineKm returns the great-circle distance between two WGS84 points
// in kilometres. Inputs and the return value are in degrees/km, not
// radians - callers should never need math.Pi in sight.
func HaversineKm(lat1, lon1, lat2, lon2 float64) float64 {
	phi1, phi2 := toRad(lat1), toRad(lat2)
	dPhi := toRad(lat2 - lat1)
	dLambda := toRad(lon2 - lon1)

	a := math.Sin(dPhi/2)*math.Sin(dPhi/2) +
		math.Cos(phi1)*math.Cos(phi2)*math.Sin(dLambda/2)*math.Sin(dLambda/2)
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	return EarthRadiusKm * c
}

// InitialBearingDeg returns the compass bearing (0-360, 0 = true north)
// of the great-circle path from (lat1,lon1) to (lat2,lon2) at its start.
// This is what a vessel's COG should read while following a lane segment
// toward that point - it is not constant along a great circle, but a
// shipping-lane segment is short enough that the difference over its
// length is not worth modelling here.
func InitialBearingDeg(lat1, lon1, lat2, lon2 float64) float64 {
	phi1, phi2 := toRad(lat1), toRad(lat2)
	dLambda := toRad(lon2 - lon1)

	y := math.Sin(dLambda) * math.Cos(phi2)
	x := math.Cos(phi1)*math.Sin(phi2) - math.Sin(phi1)*math.Cos(phi2)*math.Cos(dLambda)
	theta := math.Atan2(y, x)
	return math.Mod(toDeg(theta)+360, 360)
}

// Interpolate returns the point a fraction (0=start, 1=end) of the way
// from (lat1,lon1) to (lat2,lon2).
//
// This is LINEAR interpolation in lat/lon space, not a great-circle
// interpolation (slerp). Deliberate simplification: a shipping-lane
// segment here spans at most a few hundred km, where the two differ by
// metres - immaterial next to the AIS position noise and lane-width
// jitter the caller adds on top. Do not reuse this for anything spanning
// ocean-basin distances or crossing the antimeridian; neither is a case
// any lane in this package's caller (services/aisgen) needs.
func Interpolate(lat1, lon1, lat2, lon2, frac float64) (lat, lon float64) {
	return lat1 + (lat2-lat1)*frac, lon1 + (lon2-lon1)*frac
}
